Seeking in Netflix triggers unbounded growth in WebCore::Audio&VideoTrackPrivateMediaSourceAVFObjC objects

youssefsoli opened the pull request:
#### ce149547adc310ce20758ebfd0c37440746c232e
<pre>
Seeking in Netflix triggers unbounded growth in WebCore::Audio&amp;VideoTrackPrivateMediaSourceAVFObjC objects
<a href="https://bugs.webkit.org/show_bug.cgi?id=242925">https://bugs.webkit.org/show_bug.cgi?id=242925</a>
&lt;rdar://96300539&gt;

Reviewed by Eric Carlson.

Video/AudioTrackPrivate objects are created and stored in both the SourceBufferPrivateAVFObjC
layer and the RemoteMediaPlayerProxy layer, however, once we clear the tracks
from the SourceBufferPrivateAVFObjC layer, we fail to also delete them in
the RemoteMediaPlayerProxy layer. This causes unbounded memory growth especially
as new intialization segments + tracks keep getting added.

Instead of deleting the tracks in the HTMLMediaElement layer and send incorrect
events, we can override the &quot;old&quot; tracks in both the SourceBufferPrivateAVFObjC
and RemoteMediaPlayerProxy layers in the GPUP and still hold onto the existing
WebProcess tracks.

This patch also includes better error handling with the sourceBufferPrivateDidReceiveInitializationSegment
completion handler when a failure happens along the way.

* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::sourceBufferPrivateDidReceiveInitializationSegment):
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/AudioTrackPrivate.h:
(WebCore::AudioTrackPrivate::operator== const):
* Source/WebCore/platform/graphics/InbandTextTrackPrivate.h:
(WebCore::InbandTextTrackPrivate::operator== const):
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::didReceiveInitializationSegment):
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
* Source/WebCore/platform/graphics/SourceBufferPrivateClient.h:
(WTF::LogArgument&lt;WebCore::SourceBufferPrivateClient::ReceiveResult&gt;::toString):
* Source/WebCore/platform/graphics/SourceBufferPrivateClient.cpp: Added.
(WebCore::convertEnumerationToString):
* Source/WebCore/platform/graphics/TrackPrivateBase.cpp:
(WebCore::TrackPrivateBase::operator== const):
* Source/WebCore/platform/graphics/TrackPrivateBase.h:
* Source/WebCore/platform/graphics/VideoTrackPrivate.h:
(WebCore::VideoTrackPrivate::operator== const):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::removeAudioTrack):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::removeVideoTrack):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::removeTextTrack):
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::didParseInitializationData):
(WebCore::SourceBufferPrivateAVFObjC::clearTracks):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::clearTracks):
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
(WebCore::AppendPipeline::didReceiveInitializationSegment):
* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.cpp:
(WebCore::SourceBufferPrivateGStreamer::didReceiveInitializationSegment):
* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.h:
* Source/WebCore/platform/mock/mediasource/MockSourceBufferPrivate.cpp:
(WebCore::MockSourceBufferPrivate::didReceiveInitializationSegment):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::addRemoteAudioTrackProxy):
(WebKit::RemoteMediaPlayerProxy::addRemoteVideoTrackProxy):
(WebKit::RemoteMediaPlayerProxy::addRemoteTextTrackProxy):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateDidReceiveInitializationSegment):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateDidReceiveInitializationSegment):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.messages.in:

Canonical link: <a href="https://commits.webkit.org/252887@main">https://commits.webkit.org/252887@main</a>
</pre>

jyavenard reviewed (changes requested):
I don't think that's right.

Per spec, when a new init segment is received, and a new track is found, it's added. But previous tracks aren't removed, they marked as "active is false"
https://www.w3.org/TR/media-source-2/#sourcebuffer-init-segment-received

In this case, the new init segment will be identical to the previously received one. I think we should detect that the new init segment received is identical to the previous one and ignore it.

Otherwise we will fire events that we shouldn't fire: there are no new tracks added.

eric-carlson commented on Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
```diff
@@ -554,7 +554,7 @@ void AppendPipeline::didReceiveInitializationSegment()
     GST_DEBUG("Notifying SourceBuffer of initialization segment.");
     GST_DEBUG_BIN_TO_DOT_FILE_WITH_TS(GST_BIN(m_pipeline.get()), GST_DEBUG_GRAPH_SHOW_ALL, "append-pipeline-received-init-segment");
-    m_sourceBufferPrivate.didReceiveInitializationSegment(WTFMove(initializationSegment), []() { });
+    m_sourceBufferPrivate.didReceiveInitializationSegment(WTFMove(initializationSegment), [](bool) { });
```
It isn’t at all clear why the completions handler’s parameter is for. I think it would be better to define and use an enum than a bool

philn commented on Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
```diff
@@ -682,10 +682,10 @@ void SourceBuffer::setActive(bool active)
 }
 
-void SourceBuffer::sourceBufferPrivateDidReceiveInitializationSegment(InitializationSegment&& segment, CompletionHandler<void()>&& completionHandler)
+void SourceBuffer::sourceBufferPrivateDidReceiveInitializationSegment(InitializationSegment&& segment, CompletionHandler<void(RecieveResult)>&& completionHandler)
```
Not arguing on the name itself, but it is misspelled.

youssefsoli replied:
Nice catch!

eric-carlson reviewed (approved):
r=me

youssefsoli committed:
Seeking in Netflix triggers unbounded growth in WebCore::Audio&VideoTrackPrivateMediaSourceAVFObjC objects
https://bugs.webkit.org/show_bug.cgi?id=242925
<rdar://96300539>

Reviewed by Eric Carlson.

Video/AudioTrackPrivate objects are created and stored in both the SourceBufferPrivateAVFObjC
layer and the RemoteMediaPlayerProxy layer, however, once we clear the tracks
from the SourceBufferPrivateAVFObjC layer, we fail to also delete them in
the RemoteMediaPlayerProxy layer. This causes unbounded memory growth especially
as new intialization segments + tracks keep getting added.

Instead of deleting the tracks in the HTMLMediaElement layer and send incorrect
events, we can override the "old" tracks in both the SourceBufferPrivateAVFObjC
and RemoteMediaPlayerProxy layers in the GPUP and still hold onto the existing
WebProcess tracks.

This patch also includes better error handling with the sourceBufferPrivateDidReceiveInitializationSegment
completion handler when a failure happens along the way.

* Source/WebCore/Modules/mediasource/SourceBuffer.cpp:
(WebCore::SourceBuffer::sourceBufferPrivateDidReceiveInitializationSegment):
* Source/WebCore/Modules/mediasource/SourceBuffer.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/AudioTrackPrivate.h:
(WebCore::AudioTrackPrivate::operator== const):
* Source/WebCore/platform/graphics/InbandTextTrackPrivate.h:
(WebCore::InbandTextTrackPrivate::operator== const):
* Source/WebCore/platform/graphics/SourceBufferPrivate.cpp:
(WebCore::SourceBufferPrivate::didReceiveInitializationSegment):
* Source/WebCore/platform/graphics/SourceBufferPrivate.h:
* Source/WebCore/platform/graphics/SourceBufferPrivateClient.h:
(WTF::LogArgument<WebCore::SourceBufferPrivateClient::ReceiveResult>::toString):
* Source/WebCore/platform/graphics/SourceBufferPrivateClient.cpp: Added.
(WebCore::convertEnumerationToString):
* Source/WebCore/platform/graphics/TrackPrivateBase.cpp:
(WebCore::TrackPrivateBase::operator== const):
* Source/WebCore/platform/graphics/TrackPrivateBase.h:
* Source/WebCore/platform/graphics/VideoTrackPrivate.h:
(WebCore::VideoTrackPrivate::operator== const):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::removeAudioTrack):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::removeVideoTrack):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::removeTextTrack):
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.h:
* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::didParseInitializationData):
(WebCore::SourceBufferPrivateAVFObjC::clearTracks):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::clearTracks):
* Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp:
(WebCore::AppendPipeline::didReceiveInitializationSegment):
* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.cpp:
(WebCore::SourceBufferPrivateGStreamer::didReceiveInitializationSegment):
* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.h:
* Source/WebCore/platform/mock/mediasource/MockSourceBufferPrivate.cpp:
(WebCore::MockSourceBufferPrivate::didReceiveInitializationSegment):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::addRemoteAudioTrackProxy):
(WebKit::RemoteMediaPlayerProxy::addRemoteVideoTrackProxy):
(WebKit::RemoteMediaPlayerProxy::addRemoteTextTrackProxy):
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.cpp:
(WebKit::RemoteSourceBufferProxy::sourceBufferPrivateDidReceiveInitializationSegment):
* Source/WebKit/GPUProcess/media/RemoteSourceBufferProxy.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.cpp:
(WebKit::SourceBufferPrivateRemote::sourceBufferPrivateDidReceiveInitializationSegment):
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.h:
* Source/WebKit/WebProcess/GPU/media/SourceBufferPrivateRemote.messages.in:

Canonical link: https://commits.webkit.org/252887@main
